### ISSUE TITLE
Kubeadm will do preflight check over wrong pod-infra-container-image registry

### ISSUE THREAD

**Issue by kigawas** [kubernetes/kubeadm#1610]
TL;DR: You can work it around by `kubeadm init --config /tmp/kubeadm.yml --ignore-preflight-errors ImagePull`

## Versions

**kubeadm version** (use `kubeadm version`):
`kubeadm version: &version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:41:54Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}`

NOTICE: Custom built k8s against `release-1.14` branch, only changed the in-tree gcp provider. Diff is [here](https://github.com/kubernetes/kubernetes/compare/release-1.14...kigawas:release-1.14) 

**Environment**:
- **Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:44:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.3-beta.0.38+b22fbe605b8444-dirty", GitCommit:"b22fbe605b8444c21f3a0eaf279253991740a7b6", GitTreeState:"dirty", BuildDate:"2019-06-11T02:46:15Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
- **Cloud provider or hardware configuration**:
Not related
- **OS** (e.g. from /etc/os-release):
Not related
- **Kernel** (e.g. `uname -a`):
Not related
- **Others**:

`kubeadm init` config file:
```yaml
apiVersion: kubeadm.k8s.io/v1beta1
kind: InitConfiguration
nodeRegistration:
  kubeletExtraArgs:
    cloud-provider: gce
    cloud-config: /etc/kubernetes/cloud-config
    pod-infra-container-image: k8s.gcr.io/pause:3.1  # <- official image
---
apiVersion: kubeadm.k8s.io/v1beta1
kind: ClusterConfiguration
imageRepository: PRIVATE_REGISTRY # <- private registry
kubernetesVersion: v1.14.3-beta.0.38+b22fbe605b8444
etcd:
  local:
    imageRepository: k8s.gcr.io
dns:
  imageRepository: k8s.gcr.io
networking:
  podSubnet: 192.168.0.0/16
apiServer:
...
controllerManager:
...
```

`docker images`
```
...
PRIVATE_REGISTRY/pause                     3.1                                da86e6ba6ca1        18 months ago       742kB
k8s.gcr.io/pause                           3.1                                da86e6ba6ca1        18 months ago       742kB
```

## What happened?
1. specify official `pod-infra-container-image` in `InitConfiguration`
2. specify private `imageRepository` in `ClusterConfiguration`
3. error when `kubeadm init` if `pause` is not in PRIVATE_REGISTRY

Despite `pod-infra-container-image` specified as the official registry `k8s.gcr.io`, the preflight check is performed over the PRIVATE_REGISTRY. 

After `pause:3.1` uploaded to PRIVATE_REGISTRY, the check went fine and `pause:3.1` was pulled from **both `k8s.gcr.io` and PRIVATE_REGISTRY**, and started with the `k8s.gcr.io`'s version.

```
error execution phase preflight: [preflight] Some fatal errors occurred:
	[ERROR ImagePull]: failed to pull image PRIVATE_REGISTRY/pause:3.1: output: Error response from daemon: manifest for PRIVATE_REGISTRY/pause:3.1 not found
, error: exit status 1
[preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
```

## What you expected to happen?
No error.

## How to reproduce it (as minimally and precisely as possible)?
See above.

## Anything else we need to know?
`--ignore-preflight-errors` will work around, but it's anti-pattern.

Not sure, but it may be related to #1003

**Comment by neolit123** [comment]
this is by design.
if you are using a private registry kubeadm assumes that you have also uploaded the pause image to it.

> -infra-container-image: k8s.gcr.io/pause:3.1  # <- official image

to prevent the kubelet from downloading the pause image from `k8s.gcr.io/pause:3.1` which is the default value, you should do:

> -infra-container-image: PRIVATE_REGISTRY/pause:3.1

**Comment by kigawas** [comment]
@neolit123 
Thanks.
In my case, I do **not** want it to be download from PRIVATE_REGISTRY. I want it to be downloaded from `k8s.gcr.io`.

**Comment by neolit123** [comment]
normally users of airgapped environments upload the pause image to the same private registry to manage a pre-pull and don't rely on pulling from the internet.

an immediate workaround is to add any pause image to your private registry to satisfy the kubeadm preflight, but pass a different one in the kubelet config.

another issue here is that the `pod-infra-container-image` field is dockershim specific and might be becoming a NO-OP in the future, so i think we will be hesitant about adding extra logic to manage it.

**Comment by kigawas** [comment]
> normally users of airgapped environments upload the pause image to the same private registry to manage a pre-pull and don't rely on pulling from the internet.
> 
> an immediate workaround is to add any pause image to your private registry to satisfy the kubeadm preflight, but pass a different one in the kubelet config.
> 
> another issue here is that the `pod-infra-container-image` field is dockershim specific and might be becoming a NO-OP in the future, so i think we will be hesitant about adding extra logic to manage it.

This is what I'm doing, but it's kinda weird that it check against one registry and download from another, in my opinion.

No op in the future is another story. I'm afraid now there are not a few users encountering with this gotcha.

**Comment by neolit123** [comment]
closing as "by-design".
